bpf/tests/ipsec_redirect_tunnel.c

```diff
@@ -0,0 +1,191 @@
+/* Copyright Authors of Cilium */
+
+#define TUNNEL_MODE
+
+#include "ipsec_redirect_generic.h"
+
+#include "node_config.h"
+
+/* must define `HAVE_ENCAP 1` before including 'lib/encrypt.h'.
+ * lib/encrypt.h eventually imports overloadable_skb.h which exposes
+ * ctx_is_overlay and ctx_is_overlay_encrypted, utilized within
+ * 'ipsec_maybe_redirect_to_encrypt'
+ */
+#define HAVE_ENCAP 1
+#include "lib/encrypt.h"
+
+#include "tests/lib/ipcache.h"
+
+static __always_inline
+int ipsec_redirect_checks(__maybe_unused struct __ctx_buff *ctx, bool outer_ipv4)
+{
+	test_init();
+
+	int ret = 0;
+	__be16 proto = outer_ipv4 ? bpf_htons(ETH_P_IP) : bpf_htons(ETH_P_IPV6);
+
+	if (!outer_ipv4)
+		test_fatal("IPv6 Overlay is not yet supported.");
+
+	/*
+	 * Fill the nodemap with the destination node's ID keyed by the
+	 * destination node's IP. This is the case for VXLAN traffic where
+	 * the outer ipv4.dst == DST_NODE_IP
+	 */
+	struct node_key key = {};
+	struct node_value val = {
+		.id = DST_NODE_ID,
+		.spi = TARGET_SPI,
+	};
+
+	if (outer_ipv4) {
+		key.family = ENDPOINT_KEY_IPV4;
+		key.ip4 = DST_NODE_IP;
+	} else {
+		key.family = ENDPOINT_KEY_IPV6;
+		key.ip6 = *(union v6addr *)DST_NODE_IP_6;
+	}
+
+	map_update_elem(&cilium_node_map_v2, &key, &val, BPF_ANY);
+
+	struct encrypt_config cfg = {
+		.encrypt_key = BAD_SPI,
+	};
+	map_update_elem(&cilium_encrypt_state, &ret, &cfg, BPF_ANY);
+
+	/* Ensure we simply return from 'ipsec_maybe_redirect_to_encrypt' if
+	 * the 'MARK_MAGIC_OVERLAY_ENCRYPTED' mark is set.
+	 */
+	TEST("overlay-encrypted-mark-set", {
+		ctx->mark = MARK_MAGIC_OVERLAY_ENCRYPTED;
+		ret = ipsec_maybe_redirect_to_encrypt(ctx, proto, SOURCE_IDENTITY);
+		assert(ret == CTX_ACT_OK);
+	})
+
+	/*
+	 * NOTE1: the passedin SOURCE_IDENTITY does not matter here,
+	 * once the function determines the packet is MARK_MAGIC_OVERLAY
+	 * it will pull the identity directly from ctx.
+	 *
+	 * NOTE2: in each test we change the dst identity in the ipcache for
+	 * consistency, but it doesn't matter here, for the same reason as above.
+	 */
+
+	if (outer_ipv4)
+		ipcache_v4_add_entry(DST_IP, 0,
+				     DST_IDENTITY, DST_NODE_IP, TARGET_SPI);
+	else
+		ipcache_v6_add_entry((const union v6addr *)DST_IP_6, 0,
+				     DST_IDENTITY, DST_NODE_IP, TARGET_SPI);
+
+	/*
+	 * Ensure encryption mark is set for host-to-pod traffic and
+	 * CTX_ACT_REDIRECT is set.
+	 */
+	TEST("overlay-host-to-pod", {
+		ctx->mark = MARK_MAGIC_OVERLAY | (HOST_ID << 16);
+		ret = ipsec_maybe_redirect_to_encrypt(ctx, proto, HOST_ID);
+		assert(ctx->mark == TARGET_MARK);
+		assert(ret == CTX_ACT_REDIRECT);
+	})
+
+	/*
+	 * Ensure encryption mark is set for world-to-pod traffic and
+	 * CTX_ACT_REDIRECT is set.
+	 */
+	TEST("overlay-world-to-pod", {
+		ctx->mark = MARK_MAGIC_OVERLAY | (WORLD_ID << 16);
+		ret = ipsec_maybe_redirect_to_encrypt(ctx, proto, WORLD_ID);
+		assert(ctx->mark == TARGET_MARK);
+		assert(ret == CTX_ACT_REDIRECT);
+	})
+
```

Comment:
To be sure - is this a real scenario? I'm surprised we would be sending out overlay traffic that contains traffic from `WORLD_ID` ... but hmm, maybe `DSR-Geneve`?

Reply:
I think it's possible as for the DSR-Geneve case, but I'm not sure in that case we would hit such codepath 🤔 
With MARK_MAGIC_OVERLAY set, we never check the carried identity. The aim of the test was to make sure that it would pass independently with CTX_ACT_REDIRECT.
I've changed it accordingly to test just 1 case (pod SOURCE_IDENTITY), while leaving the exploration of whether the type of traffic you're referring to for another PR/future work.

Any feedback is appreciated, and in case feel free to solve this conversation 😃

Reply:
Ack, let's start with this and expand as needed :)